src/native/s3select/s3select_napi.cpp

```diff
@@ -0,0 +1,289 @@
+    }
+
+    void OnError(Napi::Error const& error) override
+    {
+        Napi::Env env = Env();
+        auto obj = error.Value();
+        _deferred.Reject(obj);
+    }
+};
+
+class S3SelectNapi : public Napi::ObjectWrap<S3SelectNapi>
+{
+
+public:
+    static Napi::Object Init(Napi::Env env, Napi::Object exports);
+    S3SelectNapi(const Napi::CallbackInfo& info);
+    Napi::Value Write(const Napi::CallbackInfo& info);
+    Napi::Value Flush(const Napi::CallbackInfo& info);
+    ~S3SelectNapi();
+
+private:
+    static Napi::FunctionReference constructor;
+    InputFormat input_format;
+    s3selectEngine::s3select s3select;
+    s3selectEngine::csv_object* csv_object = nullptr;
+    s3selectEngine::json_object* json_object = nullptr;
+    const uint8_t* headers_buf;
+    uint32_t headers_len;
+};
+
+Napi::Value
+S3SelectNapi::Write(const Napi::CallbackInfo& info)
+{
+    Napi::Buffer<char> buffer = info[0].As<Napi::Buffer<char>>();
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        buffer.Data(),
+        buffer.Length(),
+        false,
+        input_format,
+        headers_buf,
+        headers_len);
+    worker->Queue();
+    return worker->_deferred.Promise();
+}
+
+Napi::Value
+S3SelectNapi::Flush(const Napi::CallbackInfo& info)
+{
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        nullptr, /*No buffer for flush*/
+        0,
+        true,
+        input_format,
+        headers_buf,
+        headers_len);
+    worker->Queue();
+    return worker->_deferred.Promise();
+}
+
+Napi::FunctionReference S3SelectNapi::constructor;
+
+Napi::Object
+S3SelectNapi::Init(Napi::Env env, Napi::Object exports)
+{
+    Napi::HandleScope scope(env);
+
+    Napi::Function func = DefineClass(
+        env,
+        "S3SelectNapi",
+        {
+            InstanceMethod("write", &S3SelectNapi::Write),
+            InstanceMethod("flush", &S3SelectNapi::Flush)
+        }
+    ); //end of DefineClass
+
+    constructor = Napi::Persistent(func);
+    constructor.SuppressDestruct();
+
+    exports.Set("S3Select", func);
+    return exports;
+}
+
+void
+s3select_napi(Napi::Env env, Napi::Object exports)
+{
+    S3SelectNapi::Init(env, exports);
+}
+
+std::string
+GetStringWithDefault(Napi::Object obj, std::string name, std::string dfault)
+{
+    if (obj.Has(name)) {
+        return obj.Get(name).ToString().Utf8Value();
+    }
+    return dfault;
+}
+
+S3SelectNapi::S3SelectNapi(const Napi::CallbackInfo& info)
+    : Napi::ObjectWrap<S3SelectNapi>(info)
+{
+    Napi::Object context = info[0].As<Napi::Object>();
+    Napi::Object input_serialization_format = context.Get("input_serialization_format").As<Napi::Object>();
+    Napi::Buffer headers_buf_obj = context.Get("records_header_buf").As<Napi::Buffer<uint8_t>>();
+    headers_buf = headers_buf_obj.Data();
+    headers_len = headers_buf_obj.Length();
```

Comment:
Keeping the pointer to this memory is not safe because no one is aware that you are using this buffer's memory outside this ctor call. However since you pass in a static buffer then it is safe, but I wouldn't want to rely on this in here.

There are many ways to resolve this, most notably is to send this headers buffer on the write/flush calls as arguments, because there we already have an _args_ref that keeps the reference to it. But you can also do the same handling inside the S3SelectNapi object and keep its own _args_ref if you prefer.

Reply:
headers_buf is a paramter in Worker ctor, so it's in the _args_ref there.
Or did I misunderstand?

Reply:
It's not, only the args that are provided in `info` are inserted into _args_ref.

Reply:
Right. Added to Worker's _args_ref.